Add merge indicator to fork choice struct

realbigsean opened the pull request:
## Issue Addressed

- The merge P2P spec requires checking whether a block's parent state has a non-empty `ExecutionPayload`.  In order to avoid loading a parent state for every block we see in gossip, this PR adds an `is_merge_complete` `bool` to the `ProtoNode` struct
 so we can call `fork_choice.get_block(block.parent_root).is_merge_complete`

## Proposed Changes

- Adds the field described above
- Adds a database migration from schema 4 to schema 5, so we can add the new field to the `PersistedForkChoice`

## Additional Info

To facilitate the database migration I added the new field to `ProtoNode` and created a `LegacyProtoNode` struct (which doesn't include the new field) and `LegacySszContainser` struct. I initially tried to use `superstruct` to create variants with and without the field but it didn't seem worth the complexity because the old struct is only going to be used during the database migration.

ethDreamer committed:
Added Execution Payload from Rayonism Fork

paulhauner committed:
Add basic encode impl

paulhauner committed:
Add basic encode impl

paulhauner committed:
Add UnionSelector to ssz encode

paulhauner committed:
Add UnionSelector to ssz encode

paulhauner committed:
Comment-out existing Option impl

paulhauner committed:
Comment-out existing Option impl

paulhauner committed:
Add non-functional decode impl

paulhauner committed:
Add non-functional decode impl

ethDreamer committed:
Updated new Containers to match Merge Spec

paulhauner committed:
Complete Decode impl

paulhauner committed:
Complete Decode impl

paulhauner committed:
Tidy comments

paulhauner committed:
Tidy comments

paulhauner committed:
Remove option code

paulhauner committed:
Remove option code

paulhauner committed:
Add comments to union selector

paulhauner committed:
Add comments to union selector

paulhauner committed:
Bump ssz and ssz_derive versions

paulhauner committed:
Bump ssz and ssz_derive versions

paulhauner committed:
Remove tests

paulhauner committed:
Remove tests

paulhauner committed:
Freshen cargo.lock

paulhauner committed:
Freshen cargo.lock

paulhauner committed:
Freshen cargo.lock

paulhauner committed:
Freshen cargo.lock

paulhauner committed:
Remove unused union impl for ssz encoder

paulhauner committed:
Remove unused union impl for ssz encoder

paulhauner committed:
Add ssz tests

paulhauner committed:
Add ssz tests

paulhauner committed:
Add top-level patch

paulhauner committed:
Add top-level patch

paulhauner committed:
Add basic union tree hash method

paulhauner committed:
Add basic union tree hash method

paulhauner committed:
Tidy tree hash

paulhauner committed:
Tidy tree hash

paulhauner committed:
Fix bugs, bump version

paulhauner committed:
Fix bugs, bump version

paulhauner committed:
Add tree_hash_derive tests

paulhauner committed:
Add tree_hash_derive tests

paulhauner committed:
Patch ssz types

paulhauner committed:
Patch ssz types

paulhauner committed:
Update tree hash to 0.4.0

paulhauner committed:
Update tree hash to 0.4.0

paulhauner committed:
Finish tree hash tests

paulhauner committed:
Finish tree hash tests

paulhauner committed:
Apply enum_behaviour

paulhauner committed:
Apply enum_behaviour

paulhauner committed:
Bump eth2_ssz_types

paulhauner committed:
Bump eth2_ssz_types

paulhauner committed:
Bump ssz, ssz_derive versions

paulhauner committed:
Bump ssz, ssz_derive versions

paulhauner committed:
Add FourByteSszOption

paulhauner committed:
Add FourByteSszOption

paulhauner committed:
Apply ssz option to committee cache

paulhauner committed:
Apply ssz option to committee cache

ethDreamer committed:
Updated BeaconBlockBody for Merge Spec

ethDreamer committed:
Completed updating BeaconState and BeaconBlockBody

paulhauner committed:
Add add_parameterized to Encoder

paulhauner committed:
Add add_parameterized to Encoder

paulhauner committed:
Add "with" to Encode derive

paulhauner committed:
Add "with" to Encode derive

paulhauner committed:
Add "with" for Decode

paulhauner committed:
Add "with" for Decode

paulhauner committed:
Remove unused derive functions

paulhauner committed:
Remove unused derive functions

paulhauner committed:
Fix bugs in ssz_derive

paulhauner committed:
Fix bugs in ssz_derive

paulhauner committed:
Use only one line for ssz field attributes

paulhauner committed:
Use only one line for ssz field attributes

paulhauner committed:
Add legacy mod to ssz

paulhauner committed:
Add legacy mod to ssz

paulhauner committed:
Use legacy option in proto array

paulhauner committed:
Use legacy option in proto array

paulhauner committed:
Import super::* in legacy macro

paulhauner committed:
Import super::* in legacy macro

paulhauner committed:
Add enum_behaviour = transparent to misc structs

paulhauner committed:
Add enum_behaviour = transparent to misc structs

paulhauner committed:
Remove FourByteSszOption

paulhauner committed:
Remove FourByteSszOption

paulhauner committed:
Resolve old uses of Option for SSZ

paulhauner committed:
Resolve old uses of Option for SSZ

paulhauner committed:
Move derive tests

paulhauner committed:
Move derive tests

paulhauner committed:
Re-order crate publish

paulhauner committed:
Re-order crate publish

paulhauner committed:
Remove ssz_types dep from tree_hash

paulhauner committed:
Remove ssz_types dep from tree_hash

paulhauner committed:
Publish ssz before tree hash

paulhauner committed:
Publish ssz before tree hash

paulhauner committed:
Fix Vec<u8> variant decode derive issue

paulhauner committed:
Fix Vec<u8> variant decode derive issue

paulhauner committed:
Tidy comments

paulhauner committed:
Tidy comments

paulhauner committed:
Reinstate legacy option tests, tidy macro

paulhauner committed:
Reinstate legacy option tests, tidy macro

paulhauner committed:
Add docs for legacy impl

paulhauner committed:
Add docs for legacy impl

paulhauner committed:
Add more tests for legacy impl

paulhauner committed:
Add more tests for legacy impl

paulhauner committed:
Panic tree hash derive if enum has multi variants

paulhauner committed:
Panic tree hash derive if enum has multi variants

paulhauner committed:
Tidy committee cache

paulhauner committed:
Tidy committee cache

ethDreamer committed:
Modified ExecutionPayload<T> to use Transaction<T>

ethDreamer committed:
Mostly Finished Changes for beacon-chain.md

ethDreamer committed:
Merge branch 'unstable' into merge

ethDreamer committed:
Merge branch 'unstable' into merge

ethDreamer committed:
Added some things for fork-choice.md

paulhauner committed:
Remove old macro

paulhauner committed:
Add test for decoding genesis state

paulhauner committed:
Fix shim in committee cache

ethDreamer committed:
Merge remote-tracking branch 'paul/ssz-union' into merge

ethDreamer committed:
Merge branch 'unstable' into merge

ethDreamer committed:
Update to match new fork-choice.md/fork.md changes

ethDreamer committed:
ran cargo fmt

ethDreamer committed:
Added Missing Pieces in eth2_libp2p for Merge

ethDreamer committed:
fix ef test

ethDreamer committed:
Various Changes to Conform Closer to Merge Spec

michaelsproul committed:
Fix consensus, SSZ, tree hash & run merge EF tests (#2622)

* Update to v1.1.0-beta.4 (squash of #2548)

* SSZ, cached tree hash, EF tests

michaelsproul committed:
Fix consensus, SSZ, tree hash & run merge EF tests (#2622)

* Update to v1.1.0-beta.4 (squash of #2548)

* SSZ, cached tree hash, EF tests

realbigsean committed:
- Update the fork choice `ProtoNode` to include `is_merge_complete`
- Add database migration for the persisted fork choice

realbigsean committed:
- Update the fork choice `ProtoNode` to include `is_merge_complete`
- Add database migration for the persisted fork choice

realbigsean committed:
Merge branch 'merge-f2f' of https://github.com/sigp/lighthouse into add-merge-indicator-to-fork-choice-struct

realbigsean committed:
update tests

realbigsean committed:
update tests

paulhauner committed:
Fix clippy lints on merge-f2f (#2626)

* Remove unchecked arith from ssz_derive

* Address clippy lints in block_verfication

* Use safe math for is_valid_gas_limit

paulhauner committed:
Fix clippy lints on merge-f2f (#2626)

* Remove unchecked arith from ssz_derive

* Address clippy lints in block_verfication

* Use safe math for is_valid_gas_limit

realbigsean committed:
Small cleanup

realbigsean committed:
Small cleanup

realbigsean committed:
Merge branch 'merge-f2f' of https://github.com/sigp/lighthouse into add-merge-indicator-to-fork-choice-struct

realbigsean committed:
lints

realbigsean committed:
lints

michaelsproul commented on beacon_node/beacon_chain/src/schema_change.rs:
```diff
@@ -73,6 +75,28 @@ pub fn migrate_schema<T: BeaconChainTypes>(
             Ok(())
         }
+        // Migration for adding `is_merge_complete` field to the fork choice store.
+        (SchemaVersion(4), SchemaVersion(5)) => {
```
There's a schema 5 in `unstable` already for checkpoint sync, so this should be 5->6 once `merge-f2f` gets updated for the latest `unstable`

paulhauner committed:
Handle merge fork in web3signer (#2631)

paulhauner committed:
Checkout Cargo.lock from unstable (#2633)

paulhauner commented:
Rebasing onto `merge-f2f` should (hopefully) solve these merge conflicts pretty cleanly.

I suggest doing a `rebase -i merge-f2f` and then just deleting all the commits that aren't your own (i.e., anything before cbedda1)

paulhauner commented:
Whilst working on #2635, I noticed that we need access to the execution block hash for finalized block when we're [setting the head](https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/fork-choice.md#notify_forkchoice_updated). We don't have this on hand, so there's two options:

1. Load the finalized block from the DB, then read the execution payload block hash from that (I'm doing this for the time being).
1. Store the payload block hash on each block in the fork choice.

Going with (2) seems appealing. Whilst (1) only involves block read (not a state read) from the database, it's still IO that I think would be nice to avoid.

Instead of storing the `is_merge_complete` bool, we could store the `execution_payload.block_hash` and then say that `is_merge_complete = execution_payload.block_hash != Hash256::zero()`.

What are your thoughts @realbigsean? Sorry for the last minute info.

realbigsean committed:
Merge branch 'add-merge-indicator-to-fork-choice-struct' of https://github.com/realbigsean/lighthouse into add-merge-indicator-to-fork-choice-struct

realbigsean committed:
Revert "Fix clippy lints on merge-f2f (#2626)"

This reverts commit 076e624929c644020c5d7a5c9d9a72cfa2ac621b.

realbigsean committed:
Revert "Fix consensus, SSZ, tree hash & run merge EF tests (#2622)"

This reverts commit 08407b17d73b51f03a1a4daacd001e21964bff78.

realbigsean committed:
store execution block hash in fork choice rather than bool

realbigsean commented:
@paulhauner I rebased as you suggested, but I'm a little confused because I'm seeing https://github.com/sigp/lighthouse/pull/2627/commits/a7e866fe4907fbd23aec6e77b9738411e99cb0b9 show up in this PR's diff but it already should be in the `merge-f2f` branch?

I implemented (2) here, it wasn't much to change on my end and it seems reasonable to me :+1:

realbigsean commented:
Moving to #2643
